BDArmory/Control/BDModuleSurfaceAI.cs

```diff
@@ -390,27 +403,81 @@ void PilotLogic()
+                        if (hitCount == terrainAvoidanceHits.Length)
+                        {
+                            terrainAvoidanceHits = Physics.SphereCastAll(ray, terrainAlertDetectionRadius, terrainAlertThreatRange, (int)LayerMasks.Scenery);
+                            hitCount = terrainAvoidanceHits.Length;
+                        }
+                        if (hitCount > 0) // Found something. 
+                        {
+                            Vector3 alertNormal = Vector3.zero;
+                            var alertDistance = terrainAlertThreatRange;
+                            debugCollAngle = 0;
+                            debugCollDist = 999;
+                            using (var hits = terrainAvoidanceHits.Take(hitCount).GetEnumerator())
+                                while (hits.MoveNext())
+                                {
+                                    if (hits.Current.collider.gameObject.GetComponentUpwards<DestructibleBuilding>() != null) // Hit a building.
+                                    {
+                                        float testDist = hits.Current.distance;
+                                        if (testDist < alertDistance)
+                                        {
+                                            alertDistance = testDist;
+                                            debugCollDist = alertDistance;
+                                        }
+                                        var normal = hits.Current.normal;
+                                        var tempVector = Vector3.zero;
+                                        float collisionAngle = Vector3.Angle(vesselDir, -normal);
+                                        debugCollAngle = collisionAngle;
+                                        if (hits.Current.distance < (100 - (100 * Mathf.Cos(collisionAngle-90))) + (terrainAlertDetectionRadius / 2))
+                                        tempVector = Vector3.Reflect(vesselDir, normal); // assuming a 100m turning circle, crashing Vee can wait to start turn depending on approach angle
+                                        alertNormal += tempVector / (1 + alertDistance * alertDistance);
+                                        //if (Vector3.Angle(tempVector, (Vector3)dodgeVector) < 50)
+                                        //    dodgeVector = tempVector; //don't change dodgeVector if dodge angle suddenly radically shifts if e.g. the normal has gotten messed up due to one of the buttresses on the Astronaut Complex now pointing the steer direction into the building instead of away from it.
+                                    }
+                                }
+                            dodgeVector = alertNormal;
+                            if (Vector3.Dot(dodgeVector, vesselDir) > 0.91f) //Heading more or less straight at wall, set dodgeVector perpendicular to vessel, not rayHit normal
```

Comment:
`alertNormal` isn't normalised, so you'll want to normalise it if you're going to compare the dot product of it with a non-zero value.